### ISSUE TITLE
VeraCrypt on FreeBSD: "Error: mount: /dev/md0: No such file or directory"

### ISSUE THREAD

**Issue by strfry** [veracrypt/VeraCrypt#576]
This has already been posted, but the sourceforge bugtracker doesn't seem to be getting much attention:
https://sourceforge.net/p/veracrypt/tickets/263/

As the description implies, handling of memory disks has changed on recent FreeBSD versions, which hasn't been corrected in Veracrypt yet.
I haven't tested if this actually works with FAT partitions, as described by the original reporter.

Still, i need this for work, and would be willing to look into it.
I would appreciate some guidance though, where to look for the thing that went wrong here.

**Comment by strfry** [comment]
I found an answer to this problem.
Apparently, the FreeBSD backend doesn't even bother about detecting any fs types than FAT.
And even if it did, `mount -t ntfs` would not work, as ntfs read support has been removed from kernel, and we're left with ntfs-3g, which has so be specified via `-o mountprog=/usr/local/bin/ntfs-3g`.

Which works for me as a hardcoded fix (i don't actually need FAT at all in Veracrypt), but it would be nice to have a general solution.
Since coding a filesystem is a bit of a task, i wonder if there is another way to influence what is passed as `filesystemType` to `CoreFreeBSD::MountFileSystem`?

I found Veracrypt a bit hard to debug and follow because of the forking multi-process architecture. So a little advise would be appreciated, on to make a proper fix for this issue.

**Comment by andreas-becker** [comment]
@idrassi maybe you can give @strfry a little advise where to fix the issue so he can create a pull request.

**Comment by skateboardCherimoya** [comment]
I also need to mount ntfs devices on FreeBSD and would be very glad if this issue is resolved in the future.
@strfry Can you please tell me what exactly you patched in order to get this to work?

**Comment by o-alquimista** [comment]
@idrassi Reopen please. Still reproducible.

https://sourceforge.net/p/veracrypt/discussion/technical/thread/667da9e7e7/?limit=25
https://sourceforge.net/p/veracrypt/tickets/263/

I'm trying to mount an ext3 partition instead.

Is there any chance this would work on OpenBSD? Did anyone try? I'm about to.

**Comment by idrassi** [comment]
Indeed, it seems the automatic fs detection doesn't work anymore in FreeBSD.
I'm not sure about mdconfig link since VeraCrypt volume requires on-the-fly encryption/decryption while mdconfig is just for RAM disks. This seems to be more related to how filesystem is detected rather than the virtual device itself.
i'm not sure I can fix this as I work mainly on Windows, Linux and macOS support in VeraCrypt. Any hints are welcome.

**Comment by idrassi** [comment]
This was addressed by #1350 / commit ea7489b9 and has been included since VeraCrypt 1.26.14. 

Required FreeBSD packages:
- `e2fsprogs` for `blkid` and ext2/3/4 support
- `fusefs-exfat` for exFAT
- `fusefs-ntfs` for NTFS / `ntfs-3g`

If these helpers are missing, users may still see mount failures because VeraCrypt can't detect or delegate the filesystem mount properly.

I'm closing this as fixed.
